video2commons/frontend/wcqs.py

```diff
@@ -0,0 +1,148 @@
+        if retry_after_ts:
+            retry_after = int((retry_after_ts - datetime.now(timezone.utc)).total_seconds())
+            raise RuntimeError(
+                f'Too many requests, try again in {retry_after} seconds'
+            )
+
+        # Make the SPARQL request using the provided query.
+        response = self.session.get(
+            'https://commons-query.wikimedia.org/sparql',
+            params={'query': query},
+            headers={
+                'Accept': 'application/sparql-results+json',
+                'User-Agent': 'video2commons-bot/1.0 (https://video2commons.toolforge.org/)'
+            },
+            # Set-Cookie session refresh headers get sent with a 307 redirect.
+            allow_redirects=True,
+            timeout=30,
+        )
+        self._save_cookies()
+
+        # Respect the rate limit status code and headers.
+        #
+        # https://wikitech.wikimedia.org/wiki/Robot_policy#Generally_applicable_rules
+        if response.status_code == 429:
+            retry_after = response.headers.get('Retry-After') or 60
+            self._set_retry(int(retry_after))
+
+            raise RuntimeError(
+                f'Too many requests, try again in {retry_after} seconds'
+            )
+
+        # Handle other unexpected response codes.
+        content_type = response.headers.get('Content-Type')
+        if (
+            response.status_code < 200
+            or response.status_code >= 300
+            or content_type != 'application/sparql-results+json;charset=utf-8'
+        ):
+            raise RuntimeError(
+                f'Got unexpected response from SPARQL ({response.status_code}): {response.text}'
+            )
+
+        return response.json()
+
+    def _check_retry(self):
+        """Checks if we're rate limited before making SPARQL requests."""
+        retry_after = redisconnection.get('wcqs:retry-after')
+
+        if retry_after:
+            retry_after_ts = datetime.fromisoformat(retry_after)
+            if retry_after_ts > datetime.now(timezone.utc):
+                return retry_after_ts
+
+        return None
+
+    def _set_retry(self, retry_after: int):
+        """Updates retry-after value in Redis."""
+        retry_after_ts = datetime.now(timezone.utc) + timedelta(seconds=retry_after)
+
+        redisconnection.setex(
+            'wcqs:retry-after',
+            retry_after,
+            retry_after_ts.replace(tzinfo=timezone.utc).isoformat()
+        )
+
+    def _get_cookies(self) -> list[dict[str, Any]]:
+        """Retrieve cookies from Redis or the filesystem."""
+        cookies = redisconnection.get('wcqs:session')
+        if cookies:
+            return json.loads(cookies)
+
+        current_app.logger.warning('Pulling in WCQS session from file as fallback')
+        try:
+            # Fallback: Pull in cookies from file. Needed for initial setup.
+            with open('/srv/v2c/wcqs-session.json', 'r') as f:
```

Comment:
/srv/v2c is for backend only. Can you please change the path to `/data/project/video2commons/wcqs-session.json`?

Reply:
Thanks! Let's try this now :)